front-end/src/screens/Settings/setCredentials.tsx

```diff
@@ -0,0 +1,260 @@
+
+			if (!signRaw) {
+				return console.error('Signer not available');
+			}
+
+			const { data: startResult } = await setCredentialsStartMutation({
+				variables: {
+					address
+				}
+			});
+
+			const signMessage = startResult?.setCredentialsStart?.signMessage;
+
+			if (!signMessage) {
+				throw new Error('Challenge message not found');
+			}
+
+			const { signature } = await signRaw({
+				address,
+				data: stringToHex(signMessage),
+				type: 'bytes'
+			});
+
+			const { data: setResult } = await setCredentialsConfirmMutation({
+				variables: {
+					address,
+					password,
+					signature,
+					username
+				}
+			});
+
+			if (setResult?.setCredentialsConfirm?.token) {
+				handleTokenChange(setResult?.setCredentialsConfirm?.token);
+			}
+			currentUser.setUserDetailsContextState((prevState) => {
+				return {
+					...prevState,
+					username,
+					web3signup: false
+				};
+			});
+		} catch (error) {
+			setErr(error);
+		}
+	};
+
+	const dismissModal = () => {
+		setShowModal(false);
+	};
+
+	const openModal = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>):void => {
+		event.preventDefault();
+		event.stopPropagation();
+
+		setShowModal(true);
+	};
+
+	return (
+		<>
+			{showModal ?
+				<Modal
+					buttons={
+						<Form.Group>
+							<Button
+								content='Set Username Password'
+								disabled={loading}
+								primary
+								onClick={handleSubmit(handleSetCredentials)}
+							/>
+							<Button
+								content='Close'
+								secondary
+								onClick={dismissModal}
+							/>
+						</Form.Group>
+					}
+					centered
+					dimmer='inverted'
+					open
+					onClose={dismissModal}
+					size='tiny'
+					title={'Set Username password'}
+				>
+					<Form standalone={false} onSubmit={handleSubmit(handleSetCredentials)}>
+						{extensionNotFound?
+							<div className='card'>
+								<ExtensionNotDetected />
+							</div>
+							: null
+						}
+						{accountsNotFound?
+							<div className='card'>
+								<div className='text-muted'>You need at least one account in Polkadot-js extenstion to login.</div>
+								<div className='text-muted'>Please reload this page after adding accounts.</div>
+							</div>
+							: null
+						}
+						{isAccountLoading
+							?
+							<Loader text={'Requesting Web3 accounts'}/>
+							:
+							accounts.length > 0 &&
+							<>
+								<Form.Group>
+									<AccountSelectionForm
+										title='Choose linked account'
+										accounts={accounts}
+										address={address}
+										onAccountChange={onAccountChange}
+									/>
+								</Form.Group>
+							</>
+						}
+						<Form.Group>
+							<Form.Field width={16}>
+								<label>Username</label>
+								<input
+									className={errors.username ? 'error' : ''}
+									name='username'
+									placeholder='John'
+									ref={register({ maxLength:20, minLength:3, pattern: /^[A-Za-z0-9._-]*$/, required: true })}
```

Comment:
These pattern should be shared, they are duplicated.

Reply:
done